pkg/utils/apply/patch.go

```diff
@@ -0,0 +1,166 @@
+package apply
+
+import (
+	"encoding/json"
+
+	"github.com/oam-dev/kubevela/pkg/oam"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/jsonmergepatch"
+	"k8s.io/apimachinery/pkg/util/mergepatch"
+	"k8s.io/apimachinery/pkg/util/strategicpatch"
+	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var k8sScheme = runtime.NewScheme()
+var metadataAccessor = meta.NewAccessor()
+
+func init() {
+	_ = clientgoscheme.AddToScheme(k8sScheme)
+}
+
+// threeWayMergePatch creates a patch by computing a three way diff.
+func threeWayMergePatch(currentObj, modifiedObj runtime.Object) (client.Patch, error) {
+	current, err := json.Marshal(currentObj)
+	if err != nil {
+		return nil, err
+	}
+	original, err := getOriginalConfiguration(currentObj)
+	if err != nil {
+		return nil, err
+	}
+	modified, err := getModifiedConfiguration(modifiedObj, true)
+	if err != nil {
+		return nil, err
+	}
+
+	var patchType types.PatchType
+	var patchData []byte
+	var lookupPatchMeta strategicpatch.LookupPatchMeta
+
+	versionedObject, err := k8sScheme.New(currentObj.GetObjectKind().GroupVersionKind())
+	switch {
+	case runtime.IsNotRegisteredError(err):
+		// use JSONMergePatch for custom resources
+		// because StrategicMergePatch doesn't support custom resources
+		patchType = types.MergePatchType
+		preconditions := []mergepatch.PreconditionFunc{
+			mergepatch.RequireKeyUnchanged("apiVersion"),
+			mergepatch.RequireKeyUnchanged("kind"),
+			mergepatch.RequireMetadataKeyUnchanged("name")}
+		patchData, err = jsonmergepatch.CreateThreeWayJSONMergePatch(original, modified, current, preconditions...)
+		if err != nil {
+			return nil, err
+		}
+	case err != nil:
+		return nil, err
+	default:
+		// use StrategicMergePatch for K8s built-in resources
+		patchType = types.StrategicMergePatchType
+		lookupPatchMeta, err = strategicpatch.NewPatchMetaFromStruct(versionedObject)
+		if err != nil {
+			return nil, err
+		}
+		patchData, err = strategicpatch.CreateThreeWayMergePatch(original, modified, current, lookupPatchMeta, true)
```

Comment:
elegant way,  I use a little complex way that add preApply function pre-patch the field managed by trait when workload use 
```
ao = append(ao, client.ForceOwnership)
a.Patch(ctx, desired, client.Apply, ao...) 
```
patch the resource